### ISSUE TITLE
Discrepancy in Phase 2 Performance Compared to Released Checkpoint

### ISSUE THREAD

**Issue by jossalgon** [delyan-boychev/imaginet#1]
First of all, I would like to thank you and congratulate you on the work you have done. I find the SelfCon approach very interesting, as well as the results obtained in the ImagiNet benchmark.

## Main issue

I have been able to replicate the evaluation using the weights you provided and obtained similar results to those reported. However, I have not been able to replicate the same results using the training scripts.

### 1. Inconsistent results in phase 2

It seems that I am executing the first phase correctly, as the t-SNE visualization of the encoder outputs shows clear groupings between real and generated images. However, in the second phase, although I achieve 0.99 accuracy on the `annotation_test.txt` file, the performance on individual test sets such as `test_midjourney.txt` does not match the results obtained using your checkpoint — I reach a maximum of around 0.85.

I am using the same hyperparameters mentioned in your paper and the official training scripts.

### 2. `anchor` behavior during calibration

I noticed a behavior that I’m not sure is intentional, and I would like to confirm in case I’m missing something.

According to the paper, during the second calibration phase, the training includes specific transformations such as padding to 256×256, random resized crop, JPEG/WebP compression, rotation and flipping. These seem to correspond to the behavior when `anchor=True` and `train=True`.

However, when checking the implementation in `datasets.py`, I found that when `train=True`, the condition for applying `anchor` is never triggered. Instead, a different transformation (`albu_transform`) is applied, which uses 96×96 crops.

Relevant code excerpt from `__getitem__`:

```
def __getitem__(self, idx):
    ...
    if self.train == True:
        ...
        image = self.albu_transform(image=image_o)["image"]
    else:
        if self.anchor:
            ... # the transformations described in the paper are applied here
```

As a result, even if `anchor=True` is specified, the paper's augmentations are never applied when `train=True` (which is the case during calibration).

### 3. BatchNorm statistics update

The paper states: “We update the batch normalization statistics within the backbone’s residual blocks,” but I couldn't find any step in the script `train_linear_classifier_selfcon.py` that explicitly updates the BatchNorm statistics. Is this expected? Is the script up to date, or should this step be added manually?

### Phase 1 (`main_represent.py`)

```
alpha: 0
batch_size: 200
cosine: true
dataset: "imaginet"
epochs: 400
feat_dim: 128
grad_cache: true
grad_cache_chunk_size: 80
label: true
learning_rate: 0.005
method: "SelfCon"
model: "resnet50"
momentum: 0.9
multiview: false
num_workers: 16
optimizer_name: "sgd"
pretrained_imagenet: true
randaug: false
resume: ""
save_dir: "./save/representation"
save_freq: 50
seed: 1
selfcon_arch: "resnet"
selfcon_pos: "[False,True,False]"
selfcon_size: "fc"
temp: 0.07
warm: true
weakaug: false
weight_decay: 0.0001
```

### Phase 2 (`train_linear_classifier_selfcon.py`)

```
anchor: true
batch_size: 200
calibration_file: "../annotations/calibration.txt"
checkpoint: "./experiments/ckpts/main_imaginet.pth"
epochs: 5
exp_name: "linear_imaginet"
lr: 0.0001
resize: false
seed: 100
selfcon_pos: "[False,True,False]"
selfcon_size: "fc"
test_aug: false
weight_decay: 0.001
```

Thank you very much in advance for your time and support. I hope this is not a problem — my only intention is to faithfully replicate the procedure in order to extend it to new datasets and enrich the training.

**Comment by jossalgon** [comment]
Update: I just realized that I trained both phase 1 and phase 2 using `resnet50` instead of `resnet50nodown`, as specified in the paper. The reason for the confusion is that both the `ConResNet` class and the training scripts use `resnet50` as the default model name, so I didn't notice the mismatch.

I will re-run both stages with the correct architecture (`resnet50nodown`) and share the updated results as soon as possible. It will took 6 days.

Apologies for the oversight, and thank you for your understanding.

**Comment by delyan-boychev** [comment]
Hello @jossalgon,

Thanks for your interest in our work! I have checked the repository and found the following things:

1. I think this is due to that you have trained `resnet50`, but not `resnet50nodown`. There is large gap in performance because the main advantage of `resnet50nodown` is the no downsampling in the first convolutional layer of the network. Try retraining with the right architecture.

2. I have found out that the training classifier script on the calibration set is not updated for the `resnet50nodown` architecture. I have now committed a fix of this script. Thank you for spotting the issue!

3. In the script `train_linear_classifier_selfcon.py` this enables tuning the batch statistics (updating only mean and std):
```python
for module in model.modules():
    if isinstance(module, torch.nn.BatchNorm2d):
        module.train()
```

If you have more questions or you cannot reproduce the results, I will leave the issue open! Once again thanks for finding out the issue! If you spot anything else, let me know!

**Comment by jossalgon** [comment]
Thank you very much for your response and for reviewing the code. It looks like my mistake was indeed using the wrong architecture. I am currently retraining and will confirm as soon as it finishes. Thanks again for your time and support!

**Comment by jossalgon** [comment]
I have completed the training and can confirm that the issue was indeed caused by using `resnet50` instead of `resnet50nodown`. After retraining with the correct architecture, I was able to successfully replicate the results. Thank you very much for your help and interest!  
I would suggest setting `resnet50nodown` as the default in `main_represent.py`, but of course, feel free to decide what you think is best.

Additionally, I had to manually add back `model = model.to("cuda")` in `linear_imaginet_tb_origin_calib.py`, which had been removed in the last commit, because otherwise it raised an error. Other than that, with the latest updates, there was no need to modify `datasets.py`.

Thank you again for your time, and congratulations once more on this great work — the results are truly impressive.

Best regards, and feel free to close the issue. Thanks!

**Comment by jossalgon** [comment]
I have completed the training and can confirm that the issue was indeed caused by using `resnet50` instead of `resnet50nodown`. After retraining with the correct architecture, I was able to successfully replicate the results. Thank you very much for your help and interest!  
I would suggest setting `resnet50nodown` as the default in `main_represent.py`, but of course, feel free to decide what you think is best.

Additionally, I had to manually add back `model = model.to("cuda")` in `linear_imaginet_tb_origin_calib.py`, which had been removed in the last commit, because otherwise it raised an error. Other than that, with the latest updates, there was no need to modify `datasets.py`.

Thank you again for your time, and congratulations once more on this great work, the results are truly impressive.

Best regards, and feel free to close the issue. Thanks!

**Comment by delyan-boychev** [comment]
Thank you for your interest in our work! We really appreciate it! If you find other issues or have other questions, feel free to open another issue! I will fix the issues you mentioned!
